FEATURE: SEO Properties Generation Backend Modules

gradinarufelix opened the pull request.

gradinarufelix committed:
FEATURE: Focus Keyword Module

rolandschuetz commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -75,9 +80,64 @@ public function initializeUpdateAction(): void
+            $propertyMappingConfiguration->allowProperties('generateEmptyFocusKeywords');
+            $propertyMappingConfiguration->allowProperties('regenerateExistingFocusKeywords');
+            $propertyMappingConfiguration->allowProperties('nodeTypeFilter');
+            $propertyMappingConfiguration->allowProperties('limit');
```
allowProperties should accept multiple?

rolandschuetz commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -75,9 +80,64 @@ public function initializeUpdateAction(): void
+            $propertyMappingConfiguration->allowProperties('nodeTypeFilter');
+            $propertyMappingConfiguration->allowProperties('limit');
+        } catch (NoSuchArgumentException) {
+            // This cannot happen, otherwise we have a broken
```
Either handle the error or remove the try-catch and let the upper method handle the error

rolandschuetz commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -75,9 +80,64 @@ public function initializeUpdateAction(): void
+        return json_encode(array_map(fn (AssetModuleResultDto $item) => $item->jsonSerialize(), $resultDtos));
+    }
+
+    public function initializeGetFocusKeywordNodesAction(): void
```
protected instead of public?

gradinarufelix replied:
Die `initializeFooAction()` Methoden sind immer public.

rolandschuetz replied:
Ah, sorry is eine Action. Ja klar, muss public sein

rolandschuetz commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -75,9 +80,64 @@ public function initializeUpdateAction(): void
+                ->forProperty(PropertyMappingConfiguration::PROPERTY_PATH_PLACEHOLDER)
+                ->skipProperties('generating', 'persisting', 'persisted')
+                ->allowAllProperties();
+        } catch (NoSuchArgumentException $e) {
```
Either handle the error or remove the try-catch and let the upper method handle the error

rolandschuetz commented on Classes/Dto/AssetModuleConfigurationDto.php:
```diff
@@ -33,12 +37,13 @@ final class AssetModuleConfigurationDto
      * @param string $language
      */
-    public function __construct(bool $onlyAssetsInUse, string $propertyName, int $limit = 10, string $language = 'en')
+    public function __construct(bool $onlyAssetsInUse, string $propertyName, int $limit = 10, string $language = 'en', int $firstResult = 0)
```
The logical order seems to be $firstResult before $limit

rolandschuetz commented on Classes/Service/AssetService.php:
```diff
@@ -34,15 +33,22 @@ class AssetService
+     * @return ResultCollectionDto
      */
-    public function getAssetsThatNeedProcessing(AssetModuleConfigurationDto $configurationDto): array
+    public function getAssetsThatNeedProcessing(AssetModuleConfigurationDto $configurationDto): ResultCollectionDto
```
Ich les gerade nur den PR online, ich würde erwarten dann man eine AssetModuleConfigurationDto reingibt und eine AssetModuleResultCollectionDto raus kriegt.. oder wird die auch für andere Configurations genutzt? Dann AssetResultCollectionDto?

gradinarufelix replied:
Das ist generisch, weil die `ResultCollectionDto` nur zwei Properties hat: `items` und `nextFirstResult`. Ich sah hier keinen Grund das noch einmal weiter zu unterteilen. Falls du meinst es macht mehr Sinn das expliziter zu machen, kann ich das gerne noch ändern. 

Eventuell könnte aber `ResultCollectionDto<AssetModuleResultDto>` gehen. Ich muss mal schauen, wie das die Doctrine Collections lösen.

rolandschuetz replied:
Ja, jetzt, wo ich den Frontend-Code besser kenne, macht das so eh Sinn. 

Das Naming "getAssetsThatNeedProcessing" is etwas seltsam. Das ist einfach eine Suche basierend auf Filtern. Vielleicht einfach:
```php
public function find(AssetFilterDto $filters): ResultCollectionDto
```

rolandschuetz reviewed (commented):
Hab den Code nur online mal drüber gelesen, ein paar Ideen, vielleicht is was brauchbares dabei

rolandschuetz commented:
"Anzahl der Bilder pro Seite" müsstet "Anzahl der gleichzeitig bearbeiteten Seiten" heißen.
Hast du meine Vorschläge absichtlich nicht übernommen, findest du "Nur leere Fokus-Keywords" besser als "Ohne Fokus-Keyword"?

Error-Handling: Wenn es keinen einzigen Seitentyp mit AiPageBriefing Mixin gibt, dann sollten wir eine Anleitung anzeigen, dass man dieses Mixin hinzufügen muss.
![Bildschirmfoto 2024-04-12 um 20 15 14](https://github.com/NEOSidekick/NEOSidekick.AiAssistant/assets/735982/4660fb2d-84c0-4f96-921f-6ccd44ef9d50)

Die Setting focusKeywordGeneratorModule funktionieren alle nicht

Dimensionen-Filter hast du weggelassen, weil er zu kompliziert war?

Beim absenden kommt ein Fehler:
![Bildschirmfoto 2024-04-12 um 20 22 20](https://github.com/NEOSidekick/NEOSidekick.AiAssistant/assets/735982/4b6e596b-93cb-44c2-acab-a3bfd00b1175)

![Bildschirmfoto 2024-04-12 um 20 22 43](https://github.com/NEOSidekick/NEOSidekick.AiAssistant/assets/735982/d9f10a7c-0a6a-44b4-9e64-78bb0a3cf916)

rolandschuetz commented:
Wenn wir jetzt einen ganzen Backend-Bereich für uns haben und wir den Chat als immer unwichtiger sehen, lass und bitte ein Backend-Modul “Konfiguration” haben, und dort bette bitte “https://api.neosidekick.com/chat/account-context/edit” als iframe ein. Mach das gleich in dem PR.

rolandschuetz commented on Classes/Dto/FocusKeywordModuleConfigurationDto.php:
```diff
@@ -0,0 +1,139 @@
+    protected string $mode;
+
+    /**
+     * @var bool
```
Is there any reason to use doc comments, when the type is already defined in PHP? (really don't know, but suspect no)

gradinarufelix replied:
Ich glaube bei primitiven Typen wäre das echt egal. Bei Objekten hatten wir da aber Kompabilitätsprobleme mit Neos 7, deshalb versuche ich das generell überall im Docblock anzugeben. Doppelt gemoppelt, aber egal.

rolandschuetz replied:
Oh, was neues zu lernen. Wie verwendet Neos diese Kommentare?

gradinarufelix replied:
Das waren die DocBlocks, also das, was vor den PHP-Types üblich war. Wird für die Reflection verwendet.

rolandschuetz commented on Classes/Factory/FocusKeywordModuleResultDtoFactory.php:
```diff
@@ -0,0 +1,42 @@
+            $publicUri,
+            $node->hasProperty('title') ? $node->getProperty('title') : '',
+            $node->hasProperty('focusKeyword') ? $node->getProperty('focusKeyword') : '',
+            // todo move "language" to the configuration later
```
TODO needs to be done

rolandschuetz commented on Resources/Private/BackendModule/BackendModule.fusion:
```diff
@@ -0,0 +1,85 @@
+                action = 'nodeTypeSchema'
+                absolute = true
+                arguments = Neos.Fusion:DataStructure {
+                    # TODO: dirty hack to not have to re-implement neos:backend.configurationCacheVersion VH
```
Not really a TODO, right? So change the comment

gradinarufelix replied:
Ist copy-paste von der Neos.Ui. Würde ich ggf. drinnen lassen als Reminder bis in der UI eine bessere Lösung kommt.

rolandschuetz replied:
Dann bitte klar beschreiben, woher der Code copy-pasted is, fühlt sich sonst etwas wie ein Plagiat an. Und macht das verstehen viel schwieriger

rolandschuetz commented on Resources/Private/BackendModule/BackendModule.fusion:
```diff
@@ -0,0 +1,85 @@
+            }
+            @process.json = ${Json.stringify(value)}
+        }
+        // TODO: maybe refactor this and not use prototypes
```
TODO

rolandschuetz commented on Resources/Private/BackendModule/src/Sagas/AppSaga.ts:
```diff
@@ -52,78 +64,168 @@ export function* watchStartModule() {
+            if (e instanceof AiAssistantError) {
+                yield put(setErrorMessage(translationService.translate('NEOSidekick.AiAssistant:Error:' + e.code, e.message, {0: e.externalMessage})))
+            }
+            // todo catch other errors with a more generic message?
```
TODO

rolandschuetz commented on Classes/Factory/FocusKeywordModuleResultDtoFactory.php:
```diff
@@ -0,0 +1,42 @@
+        try {
+            $publicUri = $this->nodeLinkingService->createNodeUri($controllerContext, $node, null, 'html', true);
+        } catch (Exception|MissingActionNameException|IllegalObjectTypeException|\Neos\Flow\Property\Exception|\Neos\Flow\Security\Exception|\Neos\Neos\Exception $e) {
+            $publicUri = '';
```
How does the frontend handle this node without a uri?

rolandschuetz commented:
Hier würde ich eine Exception werfen, weil es keinen Grund gibt, hier falsche Argumente zu bekommen:
https://github.com/NEOSidekick/NEOSidekick.AiAssistant/blob/main/Classes/Service/AssetService.php#L82

Im Allgemeinen neigst du dazu viel zu viele Exceptions einfach zu schlucken (catch ohne Error-Handling). Damit kriegst du zwar schnell irgentwas zum laufen, aber du merkst programmierfehler seltener und debuggen wird viel schwerer.

Ich denke auch hier wollen wir einen Fehler werfen, da klappt ja ganz offensichtlich was nicht, und das muss der Nutzer wissen https://github.com/NEOSidekick/NEOSidekick.AiAssistant/blob/main/Classes/Service/AssetService.php#L94

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+        $workspace = $this->workspaceRepository->findByIdentifier($configurationDto->getWorkspace());
+
+        if (!$workspace) {
+            // throw
```
TODO

rolandschuetz commented:
Z.B. FocusKeywordModuleConfigurationDto, nennen Leute das wirklich im Klassennamen DTO?

rolandschuetz commented:
Zeigts das bei dir in PHP-Storm nicht an?
<img width="342" alt="Bildschirmfoto 2024-04-12 um 20 47 34" src="https://github.com/NEOSidekick/NEOSidekick.AiAssistant/assets/735982/572c2a96-2873-40b5-83ba-c5fe7f559649">
$nodeDatasThatNeedProcessing -> $nodesDataThatNeedProcessing (weil es ja mehrere Nodes sind)

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+            $nodeDatasThatNeedProcessingCount++;
+        }
+
+        $nodeDatasThatNeedProcessing = $this->reduceNodeVariantsByWorkspaces($nodeDatasThatNeedProcessing, $workspaceChain);
```
Also du iterierst als erstes selber drüber und checkst "matchFocusKeywordProperty" um dann den reduce zu machen.
Wenn jetzt in Live die Property leer is, und in meinem Workspace bereits gesetzt, würd ich dann nicht nur das aus Live kriegen?
Ich glaub du darfst erst nach dem reduce den iterate-loop machen

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+        }
+    }
+
+    protected static function matchFocusKeywordProperty(mixed $value, FocusKeywordModuleConfigurationDto $configurationDto): bool
```
Den Funktionsnamen fand ich beim ersten lesen total unklar, vielleicht focusKeywordValueMatchesSelection?

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+
+    protected static function matchFocusKeywordProperty(mixed $value, FocusKeywordModuleConfigurationDto $configurationDto): bool
+    {
+        if ($configurationDto->getMode() === 'both') {
```
Das wär ein klassiches Switch Statemeng, um klar zu machen das basierend aus dem Mode die Auswahl passiert:

```

        switch ($configurationDto->getMode()) {
            case 'both':
                return true;
            case 'only-empty':
                return empty($value);
            case 'only-existing':
                return !empty($value);
        }
```

gradinarufelix replied:
PHPStorm hat mir sogar das angeboten:

```
return match ($configurationDto->getMode()) {
    'both' => true,
    'only-empty' => empty($value),
    'only-existing' => !empty($value),
    default => false,
};
```

krsriq replied:
`match` gibt's allerdings erst mit PHP 8..

gradinarufelix replied:
Good point! @rolandschuetz Neos 7.3 geht ja mit PHP 8. Sollen wir nur mehr Neos 7.3 und PHP >= 8 erlauben?

rolandschuetz replied:
Finde ich okay, PHP 7 is eh nimma sicher https://www.php.net/supported-versions.php

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+    protected function iterate(IterableResult $iterator): Generator
+    {
+        foreach ($iterator as $object) {
+            $object = current($object);
```
$object reassignen!? Wo kommt der Code denn her?

gradinarufelix replied:
Aus den Neos-Packages.

rolandschuetz replied:
Bitte die Quelle sauber angeben.

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+        $documentNodeTypeFilter = $configurationDto->getNodeTypeFilter() ?? 'Neos.Neos:Document';
+        $mixinSubNodeTypes = $this->nodeTypeManager->getSubNodeTypes(self::MIXIN_NODE_TYPE, false);
+        $documentSubNodeTypes = $this->nodeTypeManager->getSubNodeTypes($documentNodeTypeFilter, false);
+        $intersectNodeTypes = array_intersect(array_values($mixinSubNodeTypes), array_intersect($documentSubNodeTypes));
```
Schön :)

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+     * @param array $workspaces
+     * @return QueryBuilder
+     */
+    protected function createQueryBuilder(array $workspaces): QueryBuilder
```
Ich vermute nicht von dich dann bitte im Kommentar die Quelle/Inspiration angeben, macht Updates auf neue Neos-Versionen einfacher

rolandschuetz commented on Classes/Service/NodeService.php:
```diff
@@ -0,0 +1,219 @@
+     *
+     * @return array<NodeData>
+     */
+    protected function reduceNodeVariantsByWorkspaces(array $nodes, array $workspaces): array
```
bitte im Kommentar die Quelle/Inspiration angeben, macht Updates auf neue Neos-Versionen einfacher

rolandschuetz commented on Resources/Private/BackendModule/src/index.tsx:
```diff
@@ -1,14 +1,16 @@
 import TranslationService from "./Service/TranslationService";
 import {ExternalService} from "./Service/ExternalService";
-import {omitBy, isNull} from "lodash"
+import {isNull, omitBy} from "lodash"
+import defaultModuleConfiguration from "./Util/defaultModuleConfiguration";
```
In JavaScript mixt du konstant Zeilen in denen du einen Strichpunkt verwendest, und welch wo du ihn weglässt. Bitte immer mit, so wie aktuell in der Neos UI.

rolandschuetz commented on Resources/Private/BackendModule/src/index.tsx:
```diff
@@ -45,28 +48,51 @@ document.addEventListener('DOMContentLoaded', async() => {
+    let backendMessageIdentifier
+    switch(scope) {
+        case 'altTextGeneratorModule':
+            backendMessageIdentifier = 'bulk-image-generation'; break;
```
Gibts einen grund warum backendMessageIdentifier und scope nicht einfach der gleiche Name is?

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ConfigurationForm/LimitField.tsx:
```diff
@@ -0,0 +1,33 @@
+        return (
+            <div className={'neos-control-group'}>
+                <label className={'neos-control-label'}>
+                    {this.translationService.translate('NEOSidekick.AiAssistant:AssetModule:configuration.limit.label', 'How many images per page?')}
```
translationService fehlt in den PropTypes

gradinarufelix replied:
Ist inherited von unserem eigenen `PureComponent`, welches auf dem originalen `PureComponent` von React basiert und aber common code, wie zB. die Services, bereits mitbringt. `translationService` ist ja auch nicht in `this.props` sondern direkt eine eigene Property.

rolandschuetz replied:
Ja, hab ich erst später gesehen. War nicht ganz obvious, weil imports lies man normal ja nicht. Is okay.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ConfigurationForm.tsx:
```diff
@@ -1,143 +1,52 @@
+    scope: state.app.scope,
     hasError: state.app.hasError,
-    initialConfiguration: state.app.initialModuleConfiguration,
     backendMessage: state.app.backendMessage
```
In der index holst du dir die Backend-Message, dann speicherst du es in den Store und hier holst du es wieder aus dem Store. 

Stattdessen fände ich es viel simpler zu verstehen, und leichter zum testen, wenn das alles in dieser einen Datei steht. 

Ungetestet stell ich mir das so vor:
```js
    componentDidMount() {
        const {moduleConfiguration, scope} = this.props;
        const externalService = ExternalService.getInstance();
        externalService.configure(moduleConfiguration.apiDomain, moduleConfiguration.apiKey, moduleConfiguration.userInterfaceLanguage)
        externalService.getBackendNotification(backendMessageIdentifier).then((data) => this.setState({backendMessage: data}));
    }
```

rolandschuetz commented on Resources/Private/BackendModule/src/Components/Root.tsx:
```diff
@@ -1,29 +1,34 @@
         return (
             <Provider store={store}>
                 <div className={'neos-content neos-indented neos-fluid-container'}>
                     <ErrorMessage />
```
Discussion:
Ich würde erwarten dass einen Componente ErrorMessage eine Prop Message hat und die dann rendert, aktuell unterscheidest du nicht zwischen Components die nur von außen gesteurt werden, und denen die innere Logik haben.

Ich fände es logischer den State in der App zu steurn und hier folgendes schreiben:
```{hasError ? <ErrorMessage message={errorMessage} /> : ''}```

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ItemsList.tsx:
```diff
@@ -5,26 +5,30 @@ import {AssetListItem} from "./index";
     }
 
-    assetsAsArray(): BackendAssetModuleResultDtoInterface[]
+    itemsAsArray(): StatefulModuleItem[]
```
Man liest ~9x öfter als man schreibt, deshalb optimieren wir Code immer auf Lesbarkeit. Man beginnt immer von oben nach unten zu lesen, daher wär logisch immer die hilfsfunktionen unterhalb der größeren Funktionen zu haben.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ItemsList.tsx:
```diff
@@ -5,26 +5,30 @@ import {AssetListItem} from "./index";
+        return Object.keys(items).map(key => items[key])
     }
 
     private loadingIndicator()
```
Bei React gibts das als Convention, dass alle die JSX zurückgeben mit render beginnen, also hier "renderLoadingIndicator"

Bei renderList und renderListItem machst du es, bei loadingIndicator und emptyListIndicator nicht. Bitte konsistent gleich benennen.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ItemsList.tsx:
```diff
@@ -40,20 +44,30 @@ export default class AssetList extends PureComponent {
+        switch(scope) {
+            case 'altTextGeneratorModule':
+                return <AssetListItem item={item} />;
+            case 'focusKeywordGeneratorModule':
```
Es ist ein bisschen unschön, dass du dieses Case Statement an so vielen Stellen brauchst – insgesamt 9. Denk dran, wie wir das in Fusion machen. Nur die oberste Ebenen, und das Element selber entscheidet.

Ich vermute, das müsste nur im Root und im Item selbes ein case sein, und alles andere könnte mit orchestration generisch sein. Kömma uns gerne zusammen am Montag anschauen.

In Root:
```ts
{scope == 'focusKeywordGeneratorModule' ? <ConfigurationForm intro={this.translationService.translate('NEOSidekick.AiAssistant:FocusKeywordModule:intro', '')}><FocusKeywordModuleConfigurationForm/></ConfigurationForm> : null}
```

rolandschuetz commented on Resources/Private/BackendModule/src/Components/FocusKeywordModuleConfigurationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+        dispatch(setModuleConfiguration({moduleConfiguration}))
+    }
+}))
+export default class FocusKeywordModuleConfigurationForm extends PureComponent {
```
FocusKeywordModuleConfigurationForm und AssetModuleConfigurationForm sehen für mich sehr ähnlich aus, und wiederholen immer den selben visuellen Teil, ich denk mit einer rein-visuellen SelectField-komponente (sie in Fusion Monocle) lässt sich viel Code sparen.

Eigenlich müsste ein Configuration-Array für die Filter reichen. Die Aktionen könnten dann in AssetModuleConfigurationForm auch existieren und generisch sein.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/FocusKeywordModuleConfigurationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+                {this.renderWorkspaceField()}
+                {this.renderModeField()}
+                {this.renderNodeTypeFilterField()}
+                <LimitField configuration={configuration} updateConfiguration={updateConfiguration} />
```
So umgesetzt, funktioniert LimitField nur für genau diese Konfiguration:
```<LimitField configuration={configuration} updateConfiguration={updateConfiguration} />```

Du kannst aber voll easy ein generisches Element daraus machen, und für die meisten Selects verwenden:
```<SelectField value={configuration.limit || 10} onChange={value => updateConfiguration({limit: value})} option={[5, 10, 15, 20, 25]}/>```

rolandschuetz commented on Resources/Private/BackendModule/src/Components/ItemsList.tsx:
```diff
@@ -5,26 +5,30 @@ import {AssetListItem} from "./index";
-    assets: state.app.items,
+    scope: state.app.scope,
+    items: state.app.items,
     loading: state.app.loading,
     started: state.app.started,
     hasError: state.app.hasError,
```
Du hast hier mehrere States loading, started, hasError, ... in vielen Komponenten checkst du mehrere Kombinationen davon und die Logik ist quer über viele Komponenten verteilt, soweit ich gesehen hab gibt:

 - hasError:true dann soll nur eine Fehlermeldung angezeigt werden, und der ReturnToOverviewButton
 - (!started && !hasError) dann soll das Formular angezeigt werden, und der StartModuleButton
 - (started && !hasError) dann soll die Liste angezeigt werden
    -  loading:true dann loadingIndicator 
    - (Object.keys(items).length === 0 && !loading) dann emptyListIndicator, und ReturnToOverviewButton (warum nicht zurück zum ConfigurationForm?)
    - Immer renderList (wenn hasItems, dann auch SubmitAndFetchNextButton)

Ganz kompliziert wird es z.B. beim SubmitAndFetchNextButton mit (started && hasItems && !hasError) und ich kann mir vorstellen, dass man da ganz leicht ein if falsch macht, wenn eine neue Funktion dazukommt.

Was auffällt, es gibt keine allgemeinen Buttons, also würden die innerhalb des einzelenen Falls viel einfacher zu rendern sein ohne separate condition.

 Obiges Ding erlaubt jede Menge fehlerhafter Zustände, was ist z.B. wenn started:false lnd loading:true?
 
Es gibt ein Pattern um so etwas leicht vorhersehbar zu machen, das heißt [Finite-state machine](https://en.wikipedia.org/wiki/Finite-state_machine).
Kurz gesagt, du definierst in einem Enum welche States es gibt und welche Transitions möglich sind. Daraus ergibt sich bei dir aktuell:
state: 'configure' | 'edit' | 'error'
Der LoadingState ist für das globale Ding egal und sollte nur innerhalb der Liste gesteuert werden, mit den States: 'loading' | 'empty' | 'result'

Daraus würd sich dann ergeben:
 - state=='error' dann soll nur eine Fehlermeldung angezeigt werden, und der ReturnToOverviewButton
 - state=='configure' dann soll das Formular angezeigt werden, und der StartModuleButton
 - state=='edit' dann soll die Liste angezeigt werden
    -  compontentstate = 'loading' dann loadingIndicator 
    - compontentstate = 'empty' dann emptyListIndicator, und ReturnToOverviewButton (warum nicht zurück zum ConfigurationForm?)
    - compontentstate = 'result' renderList, mit SubmitAndFetchNextButton)

Macht das für dich Sinn? Siehst du warum die if-Abfragen dadurch vorhersehbarer und einfacher werden? Falls du nicht überzeugt bist, lass uns gerne mal reden.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/SubmitAndFetchNextButton.tsx:
```diff
@@ -42,7 +42,7 @@ export default class SubmitAndFetchNextButton extends PureComponent {
                 className={'neos-button neos-button-success'}
                 disabled={isPersisting || isLoading || isGenerating || hasItemWithoutPropertyValue}>
```
Warum wollen wir bei hasItemWithoutPropertyValue das speichern nicht erlauben. Darf ich als Editor nicht bei einer Seite ganz absichtlich das Fokus-Keyword leer lassen, z.B. weil ich das nochmal mit einem Kollegen besprechen will?

gradinarufelix replied:
Das ist noch eine Condition von den Alt-Texten. Dort hast du gemeint, der Editor soll jedenfalls einen Text eingeben. Lass uns das am Montag besprechen.

rolandschuetz replied:
Ja, bei erneutem drüber nachdenken, lag ich da wahrscheinlich falsch.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/SubmitAndFetchNextButton.tsx:
```diff
@@ -42,7 +42,7 @@ export default class SubmitAndFetchNextButton extends PureComponent {
                 className={'neos-button neos-button-success'}
                 disabled={isPersisting || isLoading || isGenerating || hasItemWithoutPropertyValue}>
```
Du zeigst den Button an, während isLoading true is, ich vermute es wär logischer den hier zu verstecken!?

rolandschuetz commented on Resources/Private/BackendModule/src/Components/index.ts:
```diff
@@ -1,9 +1,9 @@
 import Root from "./Root";
-import AssetList from "./AssetList";
+import ItemsList from "./ItemsList";
 import AssetListItem from "./AssetListItem";
```
Ich glaub die Datei kann gelöscht werden!?

rolandschuetz commented on Resources/Private/BackendModule/src/Components/AssetListItem.tsx:
```diff
@@ -37,86 +36,66 @@ export default class AssetListItem extends PureComponent {
-        const {asset} = this.props;
-        return !(asset.persisted || asset.generating || asset.persisting);
+        const {item} = this.props;
+        return !(item.persisted || item.generating || item.persisting);
```
Auch das bitte als eine prop state mit fixen enum states, siehe anderer Kommentar über finite-state machines.

In dieser Komponente wirst du sehen, wie viel simpler des Code dadurck wird.

rolandschuetz commented on Resources/Private/BackendModule/src/Components/AssetListItem.tsx:
```diff
@@ -128,16 +107,16 @@ export default class AssetListItem extends PureComponent {
                             style={{marginRight: '8px'}}
                             disabled={!this.canDiscard()}
                             onClick={this.discard.bind(this)}>
```
Kleinigkeit, statt der Bind-Methode würd ich immer eine fat arrow function nutzen, also:
```onClick={() => this.discard()}>```

gradinarufelix replied:
Ist nicht dasselbe. Wenn ich das so schreibe, passiert nichts.

gradinarufelix replied:
Okay, nehme ich zurück. Es hat jetzt doch geklappt. Wenn man halt das Event benötigt, muss man es so schreiben: `(e) => this.handleChange(e)`

rolandschuetz commented on Resources/Private/BackendModule/src/Components/FocusKeywordListItem.tsx:
```diff
@@ -0,0 +1,181 @@
+        const { item, persist, generate } = this.props;
+        const textfieldId = item.identifier
+        if (item.pageContent) {
+            const iframe = this.iframeRef.current
```
Bei jedem render schreibst du den iframe inhalt neu. Ich würd vermuten, wenn du im iFrame scrollst und dann was generieren lasst springt das iFrame. Ich würd das in eine didUpdate schreiben, und zwar nur wenn sich der pageContent geändert hat.

gradinarufelix replied:
Hab es gerade probiert: wenn ich generiere oder speichere bleibt die scroll position komplett gleich. Ich nehme dann mal an, dass es keine Auswirkungen hat.

rolandschuetz commented on Resources/Private/BackendModule/src/Model/ModuleItem.ts:
```diff
@@ -0,0 +1,4 @@
+import {FocusKeywordModuleItem} from "./FocusKeywordModuleItem";
+import {AssetModuleItem} from "./AssetModuleItem";
+
+export interface ModuleItem extends AssetModuleItem, FocusKeywordModuleItem {}
```
Wenn man im BackendModule is, dann muss vermutlich nicht überall "Module" dranstehen.

rolandschuetz commented on Resources/Private/BackendModule/src/Model/ResultCollection.ts:
```diff
@@ -0,0 +1,6 @@
+import {ModuleItem} from "./ModuleItem";
+
+export interface ResultCollection {
```
Ist eigentlich kein Result, und auch keine Collection.

In sich logisch wäre denke ich das hier, inspiriert von Laravel:
```
export interface PaginagedCollection {
    items: ModuleItem[],
    currentPage: number
    perPage: number
}
```

rolandschuetz commented on Resources/Private/BackendModule/src/Sagas/AppSaga.ts:
```diff
@@ -52,78 +64,168 @@ export function* watchStartModule() {
+            console.log(propertyValue)
+            yield put(updateItemProperty({ identifier: item.identifier, propertyName, propertyValue }))
+        } catch (e) {
+            if (e instanceof AiAssistantError) {
```
Fehler die keine AiAssistantError sind sollten wir trotzdem inkl. Fehlermeldung anzeigen – wieder, wir wollen uns das debuggen einfach machen, wenn ein Nutzer ein Problem hat.

rolandschuetz commented on Resources/Private/BackendModule/src/Sagas/AppSaga.ts:
```diff
@@ -52,78 +64,168 @@ export function* watchStartModule() {
-            propertyValue: asset.propertyValue
+            identifier: item.identifier,
+            // Asset
+            filename: item.filename,
```
Wieso brauchen wir dieses mapping, was hard-coded ist, können wir nicht einfach direkt das item nutzen?

rolandschuetz commented on Resources/Private/BackendModule/src/Sagas/AppSaga.ts:
```diff
@@ -52,78 +64,168 @@ export function* watchStartModule() {
+    }
+}
+
+function *getItemGenerationConfigDependingOnScope(item: ModuleItem) {
```
Ich finde die Sagas ziemlich schwer zu lesen, kann aber auch gut sein weil ich Redux Saga nicht gscheid kenn. Das Tool ist wohl schon lange uncool, was ich so gesehen habe managen alle normal den State direkt in der Komponente. Wir könnten eine generische PaginagetItemList machen, die wir kann für jeden Use Case extenden, dann mischen wir nicht immer so komisch die zwei Fälle sondern hätten eine sauberer Kompontente in der alle Logik zusammen ist. Was denkst du?

rolandschuetz commented:
Ich vermute das [TranslationService](https://github.com/NEOSidekick/NEOSidekick.AiAssistant/blob/f8b4433f2c1b99e8dc1bf22fd95c9dec19d5615b/Resources/Private/BackendModule/src/Service/TranslationService.ts) hast du 1:1 von Neos kopiert? Dann bitte entsprechend im Code schreiben, da war doch schonmal was 😂

rolandschuetz commented on Resources/Private/BackendModule/src/Util/defaultModuleConfiguration.ts:
```diff
@@ -0,0 +1,15 @@
+export default {
```
Können wir da nicht nur die Konfiguration aus dem Backend nutzen?
Ich versuche zu vermeiden, dass man an so vielen komplett unterschiedlichen Stellen im Code etwas ändern muss, um ein neues Modul hinzuzufügen.

gradinarufelix replied:
Schlägt in dieselbe Kerbe was wir gestern am Telefon besprochen haben, wonach wir in den Settings.yaml eine Default-Konfiguration und eine Custom-Konfiguration und dann im Modul selbst eine User-Konfiguration haben. Dann wäre alles relevante in YAML.

rolandschuetz commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -75,9 +80,64 @@ public function initializeUpdateAction(): void
+                $this->arguments->getArgument('resultDtos')->getPropertyMappingConfiguration();
+            $propertyMappingConfiguration
+                ->forProperty(PropertyMappingConfiguration::PROPERTY_PATH_PLACEHOLDER)
+                ->skipProperties('generating', 'persisting', 'persisted')
```
Das sieht nach einem Coe Smell aus, warum landen diese Frontend-States überhaupt im Backend?

gradinarufelix committed:
WIP FEATURE: Focus Keyword Module - Refactoring with Roland

gradinarufelix committed:
Make language dimension name configurable

gradinarufelix committed:
Refactor components and remove outdated code

gradinarufelix committed:
WIP Refactor from store to app context

gradinarufelix commented:
> Wenn wir jetzt einen ganzen Backend-Bereich für uns haben und wir den Chat als immer unwichtiger sehen, lass und bitte ein Backend-Modul “Konfiguration” haben, und dort bette bitte “[https://api.neosidekick.com/chat/account-context/edit”](https://api.neosidekick.com/chat/account-context/edit%E2%80%9D) als iframe ein. Mach das gleich in dem PR.

@rolandschuetz @krsriq Wie hänge ich den API-Key an?

rolandschuetz committed:
TASK: Move code into component

rolandschuetz committed:
TASK: Simplify Configuration View

rolandschuetz committed:
BUGFIX: yarn package configuration, and remove redux

rolandschuetz committed:
TASK: Better error handling

gradinarufelix committed:
Add configuration module

gradinarufelix committed:
Make IDE happy and ensure better type hinting

gradinarufelix committed:
Cleanup unused files

gradinarufelix committed:
Better PHP constraint

rolandschuetz committed:
TASK: Refactor into separate components

rolandschuetz committed:
TASK: Adopt asset configuration form

gradinarufelix committed:
Remove obsolete constructor params

gradinarufelix committed:
Fix property mapping

gradinarufelix committed:
Make nodeTypeFilter optional

gradinarufelix committed:
Add temporary fallback

gradinarufelix committed:
Add temporary fallback

rolandschuetz committed:
TASK: More consistent naming

rolandschuetz committed:
WIP: Refactor backend module

rolandschuetz committed:
FEATURE: Implement focus keyword calculation

rolandschuetz committed:
WIP

rolandschuetz committed:
WIP

rolandschuetz committed:
WIP: More generalization and edge case handling

rolandschuetz committed:
TASK: Stablily sorting and pagination

rolandschuetz committed:
WIP: SEO module

rolandschuetz committed:
WIP: SEO module

rolandschuetz committed:
FEATURE: Implemented seo title and meta description generator

rolandschuetz committed:
TASK: Configurable actions

rolandschuetz committed:
TASK: Add settings schema

rolandschuetz committed:
BUGFIX: Translations

rolandschuetz committed:
Merge remote-tracking branch 'origin/main' into feature-focus-keyword-module

# Conflicts:
#	Configuration/Settings.Internal.yaml
#	Resources/Private/NeosUserInterface/src/Editors/FocusKeywordEditor.tsx
#	Resources/Private/Translations/de/BackendModule/ImageAlternativeText.xlf
#	Resources/Private/Translations/en/BackendModule/ImageAlternativeText.xlf
#	Resources/Public/NeosUserInterface/Plugin.js
#	Resources/Public/NeosUserInterface/Plugin.js.map

rolandschuetz committed:
BUGFIX: Filtering based on NodeType

rolandschuetz committed:
BUGFIX: Neos 8 compability

rolandschuetz committed:
TASK: Handle node not accessible

rolandschuetz committed:
BUGFIX: Layout overflow

rolandschuetz committed:
TASK: Feature documentation

rolandschuetz committed:
Merge remote-tracking branch 'origin/main' into feature-focus-keyword-module

# Conflicts:
#	Resources/Public/NeosUserInterface/Plugin.js
#	Resources/Public/NeosUserInterface/Plugin.js.map

rolandschuetz committed:
BUGFIX: Dont break on missing configurations

rolandschuetz committed:
TASK: Improve UX

gradinarufelix committed:
WIP BUGFIX: Get node language dimension correctly, have fallback

gradinarufelix committed:
BUGFIX: Move template to correct folder

rolandschuetz committed:
Merge remote-tracking branch 'origin/main' into feature-focus-keyword-module

rolandschuetz committed:
CHORE: Rename for consistent wording

rolandschuetz committed:
BUGFIX: Expicit excluded supertypes

rolandschuetz committed:
BUGFIX: Exclude removed and hidden not from generators

rolandschuetz committed:
TASK: Allow going to next page without changes

gradinarufelix committed:
BUGFIX: Improve Asset fetching

rolandschuetz committed:
TASK: Reduce image limit to 1000

gradinarufelix committed:
BUGFIX: Don't use node linking service to have more fine-grained control over the hostname when generating node URIs

gradinarufelix committed:
TASK: Always prepend current request domain as base uri for resources in alt text generator

gradinarufelix committed:
CLEANUP: Remove unused item

gradinarufelix committed:
Revert "BUGFIX: Don't use node linking service to have more fine-grained control over the hostname when generating node URIs"

This reverts commit 48b1284473a487a1b48872245d8d7d7a9c416063.

gradinarufelix committed:
Only return document nodes that match the current host, as to avoid errors in Fusion and CORS

gradinarufelix committed:
Split SEO properties and focus keyword property filter, to allow filtering for both separately in SEO properties generator

gradinarufelix committed:
TASK: Add language dimension filter

gradinarufelix committed:
TASK: Separate preview and public uri for document node items

gradinarufelix committed:
Allow empty default for language filter

gradinarufelix committed:
BUGFIX: Hide nodes with hidden parents

gradinarufelix committed:
Add suggestions for webandco/neos-asset-usage-cache to improve asset usage filter performance

rolandschuetz committed:
CHORE: Rename component

rolandschuetz committed:
TASK: Recommend webandco/neos-asset-usage-cache in backend UI

gradinarufelix committed:
Use new quality parameter for meta_description module

rolandschuetz committed:
CHORE: Code cleanup

gradinarufelix committed:
Implement logout and JSON error handling

gradinarufelix committed:
Server-side ordering by node path

gradinarufelix committed:
Add progress bar in list view

rolandschuetz committed:
BUGFIX: Show the text link as underline

gradinarufelix committed:
Improve sorting

rolandschuetz committed:
FEATURE: Allow multiple language selection

rolandschuetz committed:
WIP: Important pages filter and better onboarding

gradinarufelix committed:
Make progress bar nicer

gradinarufelix committed:
Implement endpoint for getting most relevant SEO uris from a given URI

gradinarufelix committed:
Design improvements

gradinarufelix committed:
BUGFIX: Re-insert new item correctly

gradinarufelix committed:
CLEANUP: Remove unused imports & build files

gradinarufelix committed:
BUGFIX: Use ServerRequest instead of Request

gradinarufelix committed:
BUGFIX: Improve error handlilng

gradinarufelix committed:
BUGFIX: Increase timeout to default value of 30s

gradinarufelix committed:
Filter important pages by empty focus keyword and missing seo title/description

gradinarufelix committed:
Fix recommendations, especially harden JSON error handling and making closures static if reasonable

gradinarufelix committed:
Do not require existing focus keywords in important pages filter for seo title/description

gradinarufelix committed:
BUGFIX: Typed class constants are only allowed since PHP 8.3

gradinarufelix committed:
Include credentials in document html fetch to omit Nginx cache

gradinarufelix committed:
Do not crash when editor configuration is missing and improve console output by adding node type and property name for better orientation

Define missing required argument "updateItemProperty" for readonly TextAreaEditor

gradinarufelix committed:
FEATURE: Show generate and reset buttons in all cases where the field is editable

gradinarufelix committed:
REFACTOR: Get fresh state when updating item to avoid overwriting intermediate changes

gradinarufelix committed:
TASK: Better handling of timeout errors when fetching most relevant pages

rolandschuetz committed:
TASK: Show plan-based backend messages

rolandschuetz committed:
BUGFIX: Better exception handling

rolandschuetz committed:
BUGFIX: Show error messages in user interface language

rolandschuetz committed:
TASK: Help the user resolve issues

rolandschuetz committed:
BUGFIX: Value must be of type number

rolandschuetz committed:
TASK: Hide placeholder on disabled focus keyword field

rolandschuetz committed:
CHORE: Improve description

rolandschuetz committed:
TASK: New build

rolandschuetz committed:
FEATURE: Show slow loading messages to user
